### ISSUE TITLE
File upload to S3 fails with file could not be saved error

### ISSUE THREAD

**Issue by rameshtek** [aws-amplify/aws-sdk-ios#2885]
Describe the bug

Some of the users had issues while uploading the file to S3.When the file is uploaded to S3,
AWS SDK writes the data to tmp file and if the write succeeds, file is uploaded to the S3.
If the file write fails then AWS SDK returns an error.

To Reproduce

Upload file to S3 by calling **uploadData** API and the API returns an error with error code 512.
This issue happens only for some users and it does not return error to all the users.Found similar issue on the 
apple developer forum.Please find the link for the same below.

https://developer.apple.com/forums/thread/128927

Please find the code snippet below.

- (AWSTask<AWSS3TransferUtilityUploadTask *> *)uploadData:(NSData *)data
                                                   bucket:(NSString *)bucket
                                                      key:(NSString *)key
                                              contentType:(NSString *)contentType
                                               expression:(AWSS3TransferUtilityUploadExpression *)expression
                                        completionHandler:(AWSS3TransferUtilityUploadCompletionHandlerBlock)completionHandler {}
 // Saves the data as a file in the temporary directory.
    NSString *fileName = [NSString stringWithFormat:@"%@.tmp", [[NSProcessInfo processInfo] globallyUniqueString]];
    NSString *filePath = [self.cacheDirectoryPath stringByAppendingPathComponent:fileName];
    NSURL *fileURL = [NSURL fileURLWithPath:filePath];
    
    NSError *error = nil;
    BOOL result = [data writeToURL:fileURL
                           options:NSDataWritingAtomic
                             error:&error];
    if (!result) {
        if (completionHandler) {
            AWSS3TransferUtilityUploadTask *uploadTask = [AWSS3TransferUtilityUploadTask new];
            uploadTask.bucket = bucket;
            uploadTask.key = key;
            completionHandler(uploadTask,error);
        }
}
Error :
Error Domain=NSCocoaErrorDomain Code=512 “The file couldn’t be saved.” UserInfo={NSUserStringVariant=Folder}

Expected behaviour
If the SDK not able to write the file to caches directory, should have an alternate location to write, so that upload 
will not fail every time. 

Environment:
Pod file versions:
'AWSCore', '2.7.0'
'AWSS3', '2.7.0'
'AWSCognito', '2.7.0'

Objective C and Swift 5

Device Information (please complete the following information):
iOS 13.5 and above
iPhone

Additional context

**Comment by palpatim** [comment]
@rameshtek The version of the SDK you are using is quite old at this point. Does this issue occur with the latest version of the SDK?

**Comment by rameshtek** [comment]
Updated the AWS SDK 2.15.0, but the issue is occurs in the update version.

**Comment by sarojkumarojha** [comment]
Hi @palpatim any update on the issue ? Thanks in advance .

**Comment by palpatim** [comment]
We'll investigate & report back when we have more info.

**Comment by palpatim** [comment]
Related to https://github.com/aws-amplify/aws-sdk-ios/issues/2884 ?

**Comment by KorolZhu** [comment]
I have the same problem of <NSCocoaErrorDomain Code=512 “The file couldn’t be saved.” UserInfo={NSUserStringVariant=Folder}>，but my app not using aws sdk, so i think is not about aws, maybe the os bug.

**Comment by ruiguoamz** [comment]
Hi, @rameshtek 
Sorry takes long to reply. Actually I have replied in issue [2884](https://github.com/aws-amplify/aws-sdk-ios/issues/2884)

1, I don't see the error on my laptop since it's really a low probability issue.
2, And as @KorolZhu said and I also took a look at the code. The error returns from Line 946 of `AWSS3TransferUtility.m` which is an Apple system call. I believe that's somehow out of our scope.
3, Just curious, the "Some of the users" you mentioned, are they seeing this issue constantly?
